src/main/java/org/springframework/data/redis/connection/jedis/JedisSubscription.java

```diff
@@ -42,11 +42,13 @@ class JedisSubscription extends AbstractSubscription {
 	 */
 	@Override
 	protected void doClose() {
-		if (!getChannels().isEmpty()) {
-			jedisPubSub.unsubscribe();
-		}
-		if (!getPatterns().isEmpty()) {
-			jedisPubSub.punsubscribe();
+		synchronized (this) {
```

Comment:
For this class, synchronization would technically not be necessary alongside my other changes. But I figure it's not a bad idea to add it in case there's a race condition somewhere else in Spring code that I've missed. 

This class is not thread safe since commands like `jedisPubSub::unsubscribe` write to the non synchronized connection `OutputStream` then flush it after. Therefore, if it's accessed by many threads, you can pollute the `OutputStream`.